src/transformers/models/llama/modeling_llama.py

```diff
@@ -1267,40 +1239,23 @@ def prepare_inputs_for_generation(
                 position_ids = position_ids[:, -input_ids.shape[1] :]
 
         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
-        if inputs_embeds is not None and past_length == 0:
+        if inputs_embeds is not None and cache_position[0] == 0:
             model_inputs = {"inputs_embeds": inputs_embeds}
         else:
-            # The `contiguous()` here is necessary to have a static stride during decoding. torchdynamo otherwise
-            # recompiles graphs as the stride of the inputs is a guard. Ref: https://github.com/huggingface/transformers/pull/29114
-            # TODO: use `next_tokens` directly instead.
+            # The `contiguous()` here is necessary to have a static stride during decoding
             model_inputs = {"input_ids": input_ids.contiguous()}
 
-        input_length = position_ids.shape[-1] if position_ids is not None else input_ids.shape[-1]
-        if cache_position is None:
-            cache_position = torch.arange(past_length, past_length + input_length, device=input_ids.device)
-        elif use_cache:
-            cache_position = cache_position[-input_length:]
```

Comment:
`generate` now always delivers the correct `cache_position`

Reply:
We still have to keep this for BC no: 
```python 
if cache_position is None:
            cache_position = torch.arange( input_length, device=input_ids.device)
```

Reply:
@ArthurZucker surprisingly no* -- these lines were right before the `return` in `prepare_inputs_for_generation`, and are the model `forward` runs these lines when `cache_position is None` there.

*the only situation that is broken is a user calling `prepare_inputs_for_generation` and then attempting to manipulate `cache_position` before passing them to `forward`. I don't think it's worth maintaining these lines for such an edge case

__________________________________________________

Important note: `cache_position` becomes a mandatory argument in this PR not because of these removed lines, but because we got rid of the `past_key_values.get_seq_length()` (which was the fallback when `cache_position is None`). 

If we want to keep `cache_position` as optional, we'll have to keep `past_key_values.get_seq_length()`. Alternativelly, we can make a deprecation cycle?

Reply:
I think cache position should always be optional, and default to input length (and not longer range of get_seq_length). 
IMO since we deprecated the usage of tuples we should also deprecate the usage of getseq length for sure!

Reply:
(But not optional for generation!)